src/java/org/apache/cassandra/repair/messages/RepairMessage.java

```diff
@@ -155,98 +153,94 @@ public static Supplier<Boolean> always()
+                RepairMetrics.retryFailure(verb);
             }
         };
-        ctx.messaging().sendWithCallback(Message.outWithFlag(verb, request, CALL_BACK_ON_FAILURE),
-                                         endpoint,
-                                         callback);
+        ctx.messaging().sendWithRetries(backoff, ctx.optionalTasks()::schedule,
+                                        verb, request, Iterators.cycle(endpoint),
+                                        (int attempt, Message<T> msg, Throwable failure) -> {
+                                            if (failure == null)
+                                            {
+                                                maybeRecordRetry.accept(attempt, null);
+                                                finalCallback.onResponse(msg);
+                                            }
+                                        },
+                                        (attempt, from, failure) -> {
+                                            ErrorHandling allowed = errorHandlingSupported(ctx, endpoint, verb, request.parentRepairSession());
+                                            switch (allowed)
+                                            {
+                                                case NONE:
+                                                    logger.error("[#{}] {} failed on {}: {}", request.parentRepairSession(), verb, from, failure);
+                                                    return false;
+                                                case TIMEOUT:
+                                                    finalCallback.onFailure(from, failure);
+                                                    return false;
+                                                case RETRY:
+                                                    if (failure == RequestFailureReason.TIMEOUT && allowRetry.get())
+                                                        return true;
+                                                    maybeRecordRetry.accept(attempt, failure);
+                                                    finalCallback.onFailure(from, failure);
+                                                    return false;
+                                                default:
+                                                    throw new AssertionError("Unknown error handler: " + allowed);
+                                            }
+                                        },
+                                        (attempt, retryReason, from, failure) -> {
+                                            switch (retryReason)
+                                            {
+                                                case MaxRetries:
+                                                    maybeRecordRetry.accept(attempt, failure);
```

Comment:
I might misunderstand, but this lambda seems to be responsible for returning an error string rather than controlling a callback. I really like how you kept this logic separate in the RemoteProcessor, maybe it's worth to do it here too?

Reply:
>  but this lambda seems to be responsible for returning an error string rather than controlling a callback

It's logging + metrics.  Logging is targeted at operators and is different based off these inputs, so are the metrics... so returning a `String` would be hard as I then need to pattern match on that string.